### ISSUE TITLE
Improve RNN Text Processing

### ISSUE THREAD

**Issue by quanhua92** [fawazsammani/The-Complete-Neural-Networks-Bootcamp-Theory-Applications#1]
Hi,
I think the code in RNN_Text Generation notebook is too complicated. You have to read the text file twice. 

```
class Dictionary(object):
    def __init__(self):
        self.word2idx = {}
        self.idx2word = {}
        self.idx = 0

    def add_word(self, word):
        if word not in self.word2idx:
            self.word2idx[word] = self.idx
            self.idx2word[self.idx] = word
            self.idx += 1
            
    def __len__(self):
        return len(self.word2idx)

class TextProcess(object):
    
    def __init__(self):
        self.dictionary = Dictionary()

    def get_data(self, path, batch_size=20):
        with open(path, 'r') as f:
            tokens = 0
            for line in f:
                words = line.split() + ['<eos>']
                tokens += len(words)
                for word in words: 
                    self.dictionary.add_word(word)  
        #Create a 1-D tensor that contains the index of all the words in the file
        rep_tensor = torch.LongTensor(tokens)
        index = 0
        with open(path, 'r') as f:
            for line in f:
                words = line.split() + ['<eos>']
                for word in words:
                    rep_tensor[index] = self.dictionary.word2idx[word]
                    index += 1
        #Find out how many batches we need            
        num_batches = rep_tensor.shape[0] // batch_size     
        #Remove the remainder (Filter out the ones that don't fit)
        rep_tensor = rep_tensor[:num_batches*batch_size]
        # return (batch_size,num_batches)
        rep_tensor = rep_tensor.view(batch_size, -1)
        return rep_tensor
```

Here is my suggestion:

```
class Dictionary(object):
    
    def __init__(self):
        self.word2idx = {}
        self.idx2word = {}
        self.idx = 0
    
    def add_word(self, word):
        if word not in self.word2idx:
            self.word2idx[word] = self.idx
            self.idx2word[self.idx] = word
            self.idx += 1
        return self.word2idx[word]
    
    def __len__(self):
        return len(self.word2idx)

class TextProcess(object):
    
    def __init__(self):
        self.dictionary = Dictionary()
        
    def get_data(self, path, batch_size=20):
        indices = []
        with open(path, 'r') as f:
            for line in f:
                words = line.split() + ['<eos>']
                for word in words:
                    idx = self.dictionary.add_word(word)
                    indices.append(idx)
                
        # Create a 1D tensor that contains the index of all the words in the file
        rep_tensor = torch.LongTensor(indices)    
        # Find out how many batches we need
        num_batches = rep_tensor.shape[0] // batch_size
        # Remove the remainder
        rep_tensor = rep_tensor[:num_batches * batch_size]
        # Return (batch_size, num_batches)
        rep_tensor = rep_tensor.view(batch_size, -1)
        return rep_tensor
```

**Comment by fawazsammani** [comment]
Hi. Yes, I wrote it long time ago. I've been working a lot with language modelling for the past year, so there is a lot to improve actually. Now i'm quite busy. I'll update the code once i'm done. Regards

**Comment by quanhua92** [comment]
@fawazsammani The training loop for RNN also needs to fix. You just used an initial zero states and didn't overwrite it with states from the model after each forward pass. Besides, you also empty the gradients and do backprop after each timestep so it just like you are training for each timestep only, no RNN at all.

Here is my modification. Please give me some feedback if I am wrong.

```python
for epoch in range(num_epochs):
    state = (torch.zeros(num_layers, batch_size, hidden_size).to(device), # hidden state
              torch.zeros(num_layers, batch_size, hidden_size).to(device)) # cell state
    
    for i in range(0, tensor.size(1) - timesteps, timesteps):
        inputs = tensor[:, i:i+timesteps]
        targets = tensor[:, (i+1):(i+1)+timesteps]
        inputs, targets = inputs.to(device), targets.to(device)    
        outputs, state = model(inputs, state)
        
    loss = criterion(outputs, targets.reshape(-1))

    model.zero_grad()
    loss.backward()

    clip_grad_norm_(model.parameters(), 0.5)
    optimizer.step()

    print(f"Epoch {epoch + 1} / {num_epochs}: Loss: {loss.item():.4f}")
```

**Comment by fawazsammani** [comment]
Hi @quanhua92. No, the implementation is correct. At the begining of each batch (each inner loop), we need to initialize the states to 0. If we replace them as in your code, then the initial hidden states of the LSTM for the next batch would be the last states of the previous batch, which is wrong. 

The code is a bit outdated. As mentioned, I will update it very soon. There are a lot better and efficient ways to implement the whole code. 

Let me know any other doubts you have.
